[clang-format] Remove special handling of comments after brace/paren

HazardyKnusperkeks opened the pull request:
Fixes http://llvm.org/PR55487 (#55487)

The code did not match the documentation about Cpp11BracedListStyle. Changed handling of comments after opening braces, which are supposedly function call like to behave exactly like their parenthesis counter part.

HazardyKnusperkeks commented:
Updated implementation and tests for `Cpp1BracedListSyle` set to `false`.

owenca commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -802,7 +802,10 @@ void ContinuationIndenter::addTokenOnCurrentLine(LineState &State, bool DryRun,
-       Previous.isOneOf(BK_BracedInit, TT_VerilogMultiLineListLParen))) {
+       (Previous.is(BK_BracedInit) &&
+        (!Style.Cpp11BracedListStyle || !Previous.Previous ||
+         Previous.Previous->isNot(tok::identifier))) ||
```
Is this inconsistent with the [documentation](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#cpp11bracedliststyle) which says, "If there is no name, a zero-length name is assumed."

owenca replied:
Yep!

HazardyKnusperkeks committed:
[clang-format] Remove special handling of comments after brace/paren

Fixes http://llvm.org/PR55487 (#55487)

The code did not match the documentation about Cpp11BracedListStyle.
Changed handling of comments after opening braces, which are supposedly
function call like to behave exactly like their parenthesis counter
part.

owenca commented on clang/unittests/Format/FormatTestComments.cpp:
```diff
@@ -4699,6 +4699,55 @@ TEST_F(FormatTestComments, SplitCommentIntroducers) {
+               Style);
+
+  verifyFormat("T baz({ // Comment\n"
+               "        arg});",
```
Why isn't this  continuation indent like others?

HazardyKnusperkeks replied:
Fixed.

owenca commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -920,12 +920,15 @@ void ContinuationIndenter::addTokenOnCurrentLine(LineState &State, bool DryRun,
-      Previous.isNot(TT_ObjCMethodExpr) && Previous.isNot(TT_RequiresClause) &&
-      Previous.isNot(TT_TableGenDAGArgOpener) &&
-      Previous.isNot(TT_TableGenDAGArgOpenerToBreak) &&
+      Previous.isNotOneOf(TT_ObjCMethodExpr, TT_RequiresClause,
```
Can we do this in an NFC patch to replace all `!isOneOf` with `isNotOneOf`?

HazardyKnusperkeks commented:
Currently containing the content of #161021, I will update once that's merged. But that should not hinder the review.

owenca commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -925,7 +925,10 @@ void ContinuationIndenter::addTokenOnCurrentLine(LineState &State, bool DryRun,
-       Previous.isOneOf(BK_BracedInit, TT_VerilogMultiLineListLParen)) &&
+       (Previous.is(BK_BracedInit) &&
+        (!Style.Cpp11BracedListStyle || !Previous.Previous ||
+         Previous.Previous->isNotOneOf(tok::identifier, tok::l_paren))) ||
```
```suggestion
         Previous.Previous->isNoneOf(tok::identifier, tok::l_paren))) ||
```

HazardyKnusperkeks replied:
This I have already done, but apparently forgot to add before the push.

owenca reviewed (commented):
Now I just realized that we should add an option for the new behavior (except questionable test cases). It seems that 5a611397a2ffdc9f1988e63a6490b0cd64267d65 only affects the formatting for `Cpp11BracedListStyle: true`, so we probably can change that option to an `enum`:
- Style1 (same as `true` before)
```c++
SomeFunction({// Comment 1
              "first entry",
              // Comment 2
              "second entry"});
```
- Style2 (space between `{` and the leading comment)
```c++
SomeFunction({ // Comment 1
               "first entry",
               // Comment 2
               "second entry"});
```
or
```c++
SomeFunction({ // Comment 1
    "first entry",
    // Comment 2
    "second entry"});
```
- Style3 (same as `false` before)
```c++
SomeFunction({ // Comment 1
               "first entry",
               // Comment 2
               "second entry" });
```
@mydeveloperday WDYT?

HazardyKnusperkeks commented on clang/unittests/Format/FormatTestComments.cpp:
```diff
@@ -4699,6 +4699,113 @@ TEST_F(FormatTestComments, SplitCommentIntroducers) {
+               Style);
+
+  verifyFormat("T baz{{ // Comment\n"
+               "        arg}};",
```
Now it is.

owenca reviewed (approved):
There are a couple of places in the documentation that mention `Cpp11BracedListStyle` being `true`, which should be updated (probably to `AlignFirstComment`).

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -4933,7 +4972,8 @@ struct FormatStyle {
   ///  This option doesn't apply to initializer braces if
-  ///  ``Cpp11BracedListStyle`` is set to ``true``.
+  ///  ``Cpp11BracedListStyle`` is set to ``AlignFirstComment`` or
+  ///  ``FunctionCall``.
```
```suggestion
  ///  ``Cpp11BracedListStyle`` is not ``Block``.
```

sstwcw commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -925,7 +925,12 @@ void ContinuationIndenter::addTokenOnCurrentLine(LineState &State, bool DryRun,
+       (Previous.is(BK_BracedInit) &&
+        (Style.Cpp11BracedListStyle != FormatStyle::BLS_FunctionCall ||
+         !Previous.Previous ||
+         Previous.Previous->isNoneOf(tok::identifier, tok::l_paren,
```
What is the part `Previous.Previous->isNoneOf(tok::identifier, tok::l_paren, BK_BracedInit)` for?

HazardyKnusperkeks replied:
If I remember correctly:
>  If the braced list follows a name (e.g. a type or variable name), clang-format formats as if the {} were the parentheses of a function call with that name. If there is no name, a zero-length name is assumed.

sstwcw replied:
I see that this piece of code is there to tell apart `x{}` from `x = {}`.  But I don't see why they should be treated differently.  Only the `FunctionCall` style has the distinction.  I thought the option `Cpp11BracedListStyle` started as a bool meaning whether or not to format all braced lists in the style that is good for C++ 11.

HazardyKnusperkeks replied:
And we only land here if `FunctionCall` is configured, otherwise `Style.Cpp11BracedListStyle != FormatStyle::BLS_FunctionCall` is true.
And before this patch `Cpp11BracedListStyle` had no effect on this whole `if`.

What's the point you want to make? It's more than half a year, I can't remember all the details or my reasoning here. It fixed the issue and aligned the (tested) implementation with the documentation, is there something wrong, or that bothers you?

sstwcw replied:
> What's the point you want to make?

I want to understand why different kinds of braced lists should be treated differently.  I am not suggesting that there is a problem with the code.

> Is there something wrong, or that bothers you?

Mostly what bothers me is that I don't understand the code.  By the way, removing this line does not break any test.

HazardyKnusperkeks replied:
There were more tests while I developed it, but I was asked to remove some of them. Maybe one of them did trigger that, but I can't completely rule out, that this check isn't needed.

owenca replied:
Dropping the conditions:
```diff
--- a/clang/lib/Format/ContinuationIndenter.cpp
+++ b/clang/lib/Format/ContinuationIndenter.cpp
@@ -926,10 +926,7 @@ void ContinuationIndenter::addTokenOnCurrentLine(LineState &State, bool DryRun,
       !(Current.MacroParent && Previous.MacroParent) &&
       (Current.isNot(TT_LineComment) ||
        (Previous.is(BK_BracedInit) &&
-        (Style.Cpp11BracedListStyle != FormatStyle::BLS_FunctionCall ||
-         !Previous.Previous ||
-         Previous.Previous->isNoneOf(tok::identifier, tok::l_paren,
-                                     BK_BracedInit))) ||
+        Style.Cpp11BracedListStyle != FormatStyle::BLS_FunctionCall) ||
        Previous.is(TT_VerilogMultiLineListLParen)) &&
       !IsInTemplateString(Current)) {
     CurrentState.Indent = State.Column + Spaces;
```
and adding back the redundant test cases from https://github.com/llvm/llvm-project/pull/71672#discussion_r2423385182 and https://github.com/llvm/llvm-project/pull/71672#discussion_r2423385357:
```diff
--- a/clang/unittests/Format/FormatTestComments.cpp
+++ b/clang/unittests/Format/FormatTestComments.cpp
@@ -4707,12 +4707,77 @@ TEST_F(FormatTestComments, LineCommentsOnStartOfFunctionCall) {
                "          value};",
                Style);
 
+  verifyFormat("T foo( // Comment\n"
+               "    arg);",
+               Style);
+
+  verifyFormat("T bar{// Comment\n"
+               "      arg};",
+               Style);
+
+  verifyFormat("T baz({// Comment\n"
+               "       arg});",
+               Style);
+
+  verifyFormat("T baz{{// Comment\n"
+               "       arg}};",
+               Style);
+
+  verifyFormat("T b0z(f( // Comment\n"
+               "    arg));",
+               Style);
+
+  verifyFormat("T b0z(F{// Comment\n"
+               "        arg});",
+               Style);
+
+  verifyFormat("func( // Comment\n"
+               "    arg);",
+               Style);
+
+  verifyFormat("func({// Comment\n"
+               "      arg});",
+               Style);
+
   Style.Cpp11BracedListStyle = FormatStyle::BLS_Block;
   verifyFormat("Type name{ // Comment\n"
                "           value\n"
                "};",
                Style);
 
+  verifyFormat("T foo( // Comment\n"
+               "    arg);",
+               Style);
+
+  verifyFormat("T bar{ // Comment\n"
+               "       arg\n"
+               "};",
+               Style);
+
+  verifyFormat("T baz({ // Comment\n"
+               "        arg });",
+               Style);
+
+  verifyFormat("T baz{ { // Comment\n"
+               "         arg } };",
+               Style);
+
+  verifyFormat("T b0z(f( // Comment\n"
+               "    arg));",
+               Style);
+
+  verifyFormat("T b0z(F{ // Comment\n"
+               "         arg });",
+               Style);
+
+  verifyFormat("func( // Comment\n"
+               "    arg);",
+               Style);
+
+  verifyFormat("func({ // Comment\n"
+               "       arg });",
+               Style);
+
   Style.Cpp11BracedListStyle = FormatStyle::BLS_FunctionCall;
   verifyFormat("Type name{ // Comment\n"
                "    value};",
```
wouldn't break any tests. So either the conditions were superfluous or we missed test cases for them.

HazardyKnusperkeks replied:
Then let's remove them: https://github.com/llvm/llvm-project/pull/191872
Most likely they handled some of the cases before I introduced the enum, but not everything and I forgot to revert that.
